felix/nftables/table.go

```diff
@@ -0,0 +1,1294 @@
+		logCxt := t.logCxt.WithField("chainName", chainName)
+		if t.dirtyChains.Contains(chainName) || t.dirtyBaseChains.Contains(chainName) {
+			// Already an update pending for this chain; no point in flagging it as
+			// out-of-sync.
+			logCxt.Debug("Skipping known-dirty chain")
+			continue
+		}
+		dpHashes := dataplaneHashes[chainName]
+		if !t.ourChainsRegexp.MatchString(chainName) {
+			// Not one of our chains, or any Calico table's chains - so it may be one that we're inserting rules into.
+			insertedRules := t.chainToInsertedRules[chainName]
+			if len(insertedRules) == 0 {
+				// This chain shouldn't have any inserts, make sure that's the
+				// case.  This case also covers the case where a chain was removed,
+				// making dpHashes nil.
+				dataplaneHasInserts := false
+				for _, hash := range dpHashes {
+					if hash != "" {
+						dataplaneHasInserts = true
+						break
+					}
+				}
+				if dataplaneHasInserts {
+					logCxt.WithField("actualRuleIDs", dpHashes).Warn(
+						"Chain had unexpected inserts, marking for resync")
+					t.dirtyBaseChains.Add(chainName)
+				}
+				continue
+			}
+
+			// Re-calculate the expected rule insertions based on the current length
+			// of the chain (since other processes may have inserted/removed rules
+			// from the chain, throwing off the numbers).
+			expectedHashes, _, _ = t.expectedHashesForInsertAppendChain(
+				chainName,
+				numEmptyStrings(dpHashes),
+			)
+			if !reflect.DeepEqual(dpHashes, expectedHashes) {
+				logCxt.WithFields(log.Fields{
+					"expectedRuleIDs": expectedHashes,
+					"actualRuleIDs":   dpHashes,
+				}).Warn("Detected out-of-sync inserts, marking for resync")
+				t.dirtyBaseChains.Add(chainName)
+			}
+		} else {
+			// One of our chains, should match exactly.
+			if !reflect.DeepEqual(dpHashes, expectedHashes) {
+				logCxt.WithFields(log.Fields{
+					"dpHashes":       dpHashes,
+					"expectedHashes": expectedHashes,
+				}).Warn("Detected out-of-sync Calico chain, marking for resync")
+				t.dirtyChains.Add(chainName)
+			}
+		}
+	}
+
+	// Now scan for chains that shouldn't be there and mark for deletion.
+	t.logCxt.Debug("Scanning for unexpected nftables chains")
+	for chainName, dataplaneHashes := range dataplaneHashes {
+		logCxt := t.logCxt.WithField("chainName", chainName)
+		if t.dirtyChains.Contains(chainName) || t.dirtyBaseChains.Contains(chainName) {
+			// Already an update pending for this chain.
+			logCxt.Debug("Skipping known-dirty chain")
+			continue
+		}
+		if _, ok := t.chainToDataplaneHashes[chainName]; ok {
+			// Chain expected, we'll have checked its contents above.
+			logCxt.Debug("Skipping expected chain")
+			continue
+		}
+		if !t.ourChainsRegexp.MatchString(chainName) {
+			// Non-calico chain that is not tracked in chainToDataplaneHashes. We
+			// haven't seen the chain before and we haven't been asked to insert
+			// anything into it.  Check that it doesn't have an rule insertions in it
```

Comment:
I think if we just always mark chains that are not ours as dirty, this code block becomes irrelevant as well.